miner/Miner.hs

```diff
@@ -59,91 +56,67 @@
 import Control.Scheduler (Comp(..), replicateWork, terminateWith, withScheduler)
-
-import Data.Aeson (ToJSON(..), Value(..))
-import qualified Data.ByteString.Char8 as B
-import qualified Data.ByteString.Base16 as B16
-import Data.Bytes.Put (runPutS)
-import qualified Data.Text as T
-import Data.Text.Encoding (encodeUtf8, decodeUtf8)
+import Data.Generics.Product.Fields (field)
 import Data.Tuple.Strict (T3(..))
-
 import Network.Connection (TLSSettings(..))
 import Network.HTTP.Client
 import Network.HTTP.Client.TLS (mkManagerSettings)
 import Network.Wai.EventSource (ServerEvent(..))
 import Network.Wai.EventSource.Streaming (withEvents)
-
 import Options.Applicative
-
+import RIO
+import RIO.List.Partial (head)
+import qualified RIO.Text as T
 import Servant.Client
-
 import qualified Streaming.Prelude as SP
-
 import qualified System.Random.MWC as MWC
 
-import Text.Pretty.Simple (pPrintNoColor)
-
 -- internal modules
 
-import Chainweb.BlockHash
-import Chainweb.BlockHeader
-import Chainweb.Difficulty
+import Chainweb.BlockHeader (Nonce(..))
 import Chainweb.HostAddress (HostAddress, hostAddressToBaseUrl)
 import Chainweb.Miner.Core
 import Chainweb.Miner.Pact (Miner, pMiner)
 import Chainweb.Miner.RestAPI.Client (solvedClient, workClient)
-import Chainweb.PowHash
-import Chainweb.Utils (runGet, textOption, toText)
+import Chainweb.Utils (textOption, toText, runGet)
 import Chainweb.Version
 
 --------------------------------------------------------------------------------
 -- CLI
 
--- | Newtype'd so that I can provide a custom `ToJSON` instance.
+-- | Result of parsing commandline flags.
 --
-newtype NodeURL = NodeURL { _url :: BaseUrl }
-
-instance Show NodeURL where
-    show (NodeURL b) = showBaseUrl b
-
-instance ToJSON NodeURL where
-    toJSON = String . T.pack . show
-
 data ClientArgs = ClientArgs
-    { cmd :: Command
-    , version :: ChainwebVersion
-    , coordinator :: NodeURL
-    , miner :: Miner
-    , chainid :: Maybe ChainId }
-    deriving stock (Show, Generic)
-    deriving anyclass (ToJSON)
+    { cmd :: !Command
+    , version :: !ChainwebVersion
+    , ll :: !LogLevel
+    , coordinator :: !BaseUrl
+    , miner :: !Miner
+    , chainid :: !(Maybe ChainId) }
+    deriving stock (Generic)
+
+-- | The top-level git-style CLI "command" which determines which mining
+-- paradigm to follow.
+--
+data Command = CPU CPUEnv | GPU
 
 newtype CPUEnv = CPUEnv { cores :: Word16 }
-    deriving stock (Show, Generic)
-    deriving anyclass (ToJSON)
-
-data GPUEnv = GPUEnv
-    deriving stock (Show, Generic)
-    deriving anyclass (ToJSON)
-
-data Command = CPU CPUEnv | GPU GPUEnv
-    deriving stock (Show, Generic)
-    deriving anyclass (ToJSON)
 
 data Env = Env
-    { gen :: MWC.GenIO
-    , mgr :: Manager
-    , args :: ClientArgs }
+    { gen :: !MWC.GenIO
+    , mgr :: !Manager
+    , log :: !LogFunc
+    , args :: !ClientArgs }
+    deriving stock (Generic)
+
+instance HasLogFunc Env where
+    logFuncL = field @"log"
```

Comment:
The only boilerplate that RIO demands for its logging.

Reply:
Please no type applications where a simple print statement would work just fine.

Reply:
Not quite the same thing. `logFuncL` is a lens, it's how the ReaderT stack gets its logging "for free".